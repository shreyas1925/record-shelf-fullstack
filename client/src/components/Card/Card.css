.grey--container .container .container--card{
    /* flex-basis:32%; */
    margin: 10px 0px;
    background-color: #DFDFDF;
    height: 240px;
    width: 225px;
    padding:15px 15px;
    border-radius: 20px;
    box-shadow: 2px 4px 12px 1px rgba(0, 0, 0.2, 0.2);
}

.grey--container .container .container--card img{
    border-radius: 50%;
}

.grey--container .container .container--card:nth-child(even)
{
    background-color: #ABABAB;
}

.container--card img{
    width: 150px;
    height: 150px;
    margin: auto;
    display: block;
}

.card--footer{
    display:flex;
    justify-content: space-between;
    margin-top: 15px;
}

.red--heart{
    display: flex;
    flex-direction: row;
    align-items: center;
    position: relative;
    cursor: pointer;
}

.red--heart img{
    width: 33px;
    height: 33px;
}

.red--heart p{
    position: absolute;
    left: 12px;
    color: white;
}

.card--details h3{
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: 1; /* number of lines to show */
            line-clamp: 1; 
    -webkit-box-orient: vertical;
}

@media screen and (max-width: 868px) {
    .grey--container .container .container--card{
        width: 300px;
        height: 300px;
        /* margin: 30px 30px; */
        margin-left: -15px;
    }

    .grey--container .container .container--card > img{
        width: 200px;
        height: 200px;
        margin: auto;
        display: block;
    }
    .red--heart img{
        width: 33px;
        height: 33px;
    }

}