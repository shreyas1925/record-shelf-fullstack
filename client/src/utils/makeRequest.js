// import axios from "axios";


// const makeRequest = async (
//   apiEndPoint,
//   dynamicConfig = {},
// ) => {
//   try{
//       const requestDetails = {
//         baseURL: BACKEND_URL,
//         url: apiEndPoint.url,
//         method: apiEndPoint.method,
//         ...dynamicConfig,
//     };
//     const { data } = await axios(requestDetails);
//     return data;
//   }catch(e)
//   {
//     console.log(e);
//   }
// };

// export default makeRequest;


const BACKEND_URL = "http://localhost:8080/";
const makeRequest = async (path, options) => {
    try{
        options.headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'Authorization': 'Bearer QWlzaHdhcnlhIE4='
        };
        const result = await fetch(`${BACKEND_URL}${path}`,options);
        const {data} = await result.json();
        return data
    } catch(err){
        return err;
    }
};

export default makeRequest;