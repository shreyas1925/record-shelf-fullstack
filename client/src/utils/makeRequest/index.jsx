
// const makeRequest = async (path, options) => {
//     try{
//         options.headers = {
//             'Content-Type': 'application/json',
//             'Accept': 'application/json',
//             'Authorization': 'Bearer QWlzaHdhcnlhIE4='
//         };
//         const result = await fetch(`${BACKEND_URL}${path}`,options);
//         const {data} = await result.json();
//         return data
//     } catch(err){
//         return err;
//     }
// };

// export default makeRequest;

import axios from 'axios';
const BACKEND_URL = 'http://localhost:8080/';
const makeRequest = async (apiEndPoint, dynamicConfig, navigate) => {
    try {
        const requestDetails = {
            baseURL: BACKEND_URL,
            url: apiEndPoint.url,
            method: apiEndPoint.method,
            headers: {
                authorization: 'Bearer QWlzaHdhcnlhIE4=',
            },
            ...dynamicConfig,
        };
        const { data } = await axios(requestDetails);
        return data;
    } catch (error) {
        const errorCode = error.response?.status;
        if (errorCode) {
            navigate(`/error/${errorCode}`);
        } else {
            navigate('/error');
        }
    }
};
  
export default makeRequest;