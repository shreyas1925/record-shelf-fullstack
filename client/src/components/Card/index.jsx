import React, {useEffect, useState } from "react";
import "./Card.css";
import redHeart from "../../assets/heart-red.svg";
import blackHeart from "../../assets/heart-gray.svg";
import makeRequest from "../../utils/makeRequest";
import { GET_SONG_LIKE_BY_ID, UPDATE_SONG_LIKE_BY_ID } from "../../constants/apiEndPoints";

const Card = ({ song }) => {
  
  const [countInc, setCountInc] = useState(0);
  const [likeInc, setLikeInc] = useState(false);

  useEffect(()=>{
    makeRequest(GET_SONG_LIKE_BY_ID(song.id), {})
        .then((response) => {
          setCountInc(response.data.count)
          setLikeInc(response.data.like)
        })
        .catch((e) => {
          // setError(e.message);
      });
  })
  
  const likedImage = likeInc ? redHeart : blackHeart;

  const handleLikes = async () => {
    try{
      await makeRequest(UPDATE_SONG_LIKE_BY_ID(song.id),{ data: { like:!likeInc } });
      setCountInc(likeInc ? countInc - 1 : countInc + 1);
      setLikeInc(!likeInc);
    } catch(e){
      // 
    }
  }
  return (
    <div className="container--card" data-testid='card'>
      <img alt="card" src={song.imageUrl} />
      <div className="card--footer">
        <div className="card--details">
          <h3>{song.name}</h3>
          <p>{song.artist.name}</p>
        </div>
        <div className="red--heart">
          <img src={likedImage} alt="heart" onClick={handleLikes}/>
          <p>{countInc}</p>
        </div>
      </div>
    </div>
  );
};

export default Card;
