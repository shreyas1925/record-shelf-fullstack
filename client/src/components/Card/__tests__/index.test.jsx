import Card from "..";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import makeRequest from "../../../utils/makeRequest";

jest.mock("../../../utils/makeRequest");

const mockData = {
  id: "45e1d753-2986-43cb-9b9d-30c370056319",
  name: "This Love",
  genre: {
    id: "128aa7f8-c943-48ce-b352-7edd26fa4c6e",
    name: "Pop",
  },
  artist: {
    id: "496b0a85-2bfa-45bc-8d0f-57fe0ce55708",
    name: "Maroon 5",
  },
  imageUrl: "https://i.scdn.co/image/ab67616d0000b27317b3850d758fff5a2301e537",
  publishedAt: "2002-06-25T00:00:00",
};

const mockConstants = {
  data: { count: 0, like: false },
};

describe("Card", () => {
  it("should render without crashing", () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const cardElement = screen.getByTestId("card");
    expect(cardElement).toBeInTheDocument();
  });

  it("should render contents", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const countElement = screen.getByText("Maroon 5");
    const altElement = screen.getByAltText("heart");
    expect(altElement.src).toContain("heart-gray.svg");
    expect(countElement).toBeInTheDocument();
  });
});

describe("Like Count", () => {
  it("should be 0 before firing event", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const countElement = screen.getByText("0");
    expect(countElement).toBeInTheDocument();
  });

  it("should be 1 after firing event once", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const countElement = screen.getByText("0");
    expect(countElement).toBeInTheDocument();
    fireEvent.click(screen.getByAltText("heart"));
    await waitFor(() => {
      const countElement = screen.getByText("1");
      expect(countElement).toBeInTheDocument();
    });
  });
  it("should be 0 when firing event done twice", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const countElement = screen.getByText("0");
    expect(countElement).toBeInTheDocument();
    fireEvent.click(screen.getByAltText("heart"));
    await waitFor(() => {
      const countElement = screen.getByText("1");
      expect(countElement).toBeInTheDocument();
    });
    fireEvent.click(screen.getByAltText("heart"));
    await waitFor(() => {
      const countElement = screen.getByText("0");
      expect(countElement).toBeInTheDocument();
    });
  });
});

describe("Like Heart", () => {
  it("should be gray heart before firing event", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const imageElement = screen.getByAltText("heart");
    expect(imageElement.src).toContain("heart-gray.svg");
  });

  it("should be red heart after firing event once", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const imageElement = screen.getByAltText("heart");
    expect(imageElement.src).toContain("heart-gray.svg");
    fireEvent.click(screen.getByAltText("heart"));
    await waitFor(() => {
      const imageElement = screen.getByAltText("heart");
      expect(imageElement.src).toContain("heart-red.svg");
    });
  });

  it("should be gray heart when firing event done twice", async () => {
    makeRequest.mockResolvedValue(mockConstants);
    render(<Card song={mockData} />);
    const imageElement = screen.getByAltText("heart");
    expect(imageElement.src).toContain("heart-gray.svg");
    fireEvent.click(screen.getByAltText("heart"));
    await waitFor(() => {
      const imageElement = screen.getByAltText("heart");
      expect(imageElement.src).toContain("heart-red.svg");
    });
    fireEvent.click(screen.getByAltText("heart"));
    await waitFor(() => {
      const imageElement = screen.getByAltText("heart");
      expect(imageElement.src).toContain("heart-red.svg");
    });
  });
});

describe("Card Snapshot", () => {
    it("should render forming a snapshot", () => {
      makeRequest.mockResolvedValue(mockConstants);
      const { asFragment } = render(<Card song={mockData} />);
      expect(asFragment()).toMatchSnapshot(); 
    });
});
