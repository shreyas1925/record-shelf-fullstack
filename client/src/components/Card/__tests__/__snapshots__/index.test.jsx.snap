// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Snapshot should render forming a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container--card"
    data-testid="card"
  >
    <img
      alt="card"
      src="https://i.scdn.co/image/ab67616d0000b27317b3850d758fff5a2301e537"
    />
    <div
      class="card--footer"
    >
      <div
        class="card--details"
      >
        <h3>
          This Love
        </h3>
        <p>
          Maroon 5
        </p>
      </div>
      <div
        class="red--heart"
      >
        <img
          alt="heart"
          src="heart-gray.svg"
        />
        <p>
          0
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
